### PR TITLE
CASSANDRA-17062 only cache refactoring

### PR DESCRIPTION
**Pull request by smiklosovic**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: src/java/org/apache/cassandra/cache/UnweightedCacheSize.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+{
+    /**
+     * Returns the maximum number of entries in this cache.
+     */
+    int maxEntries();
+
+    /**
+     * Specifies the maximum number of entries in this cache.
+     */
+    void setMaxEntries(int maxEntries);
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/cache/UnweightedCacheSize.java:39]: This method is not used.

File: src/java/org/apache/cassandra/metrics/AbstractCacheMetrics.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import com.google.common.annotations.VisibleForTesting;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.RatioGauge;
+
+import static org.apache.cassandra.metrics.CassandraMetricsRegistry.Metrics;
+
+/**
+ * Metrics for an unweighted cache like {@code AuthCache}.
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/metrics/AbstractCacheMetrics.java]: The comment is not accurate - this class is used to hold both unweighted and weighted cache metrics.

File: src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+import com.codahale.metrics.Gauge;
+import org.apache.cassandra.cache.UnweightedCacheSize;
+
+import static org.apache.cassandra.metrics.CassandraMetricsRegistry.Metrics;
+
+/**
+ * Metrics for an unweighted cache like {@code AuthCache}.
+ */
+public class UnweightedCacheMetrics extends AbstractCacheMetrics
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java:28]: It took me a while to figure out the difference between weighted and unweighted terms, so I assume the users will have to do the same exercise. At the moment,the `UnweightedCache` fully matches the `AuthCache`, so maybe we should classify these types of caches in this way, wdyt? This should be easier to understand for a user.

File: src/java/org/apache/cassandra/cache/ICache.java [hunk 4]
```diff
@@ -23,6 +23,8 @@
  * This is similar to the Map interface, but requires maintaining a given capacity
  * and does not require put or remove to return values, which lets SerializingCache
  * be more efficient by avoiding deserialize except on get.
+ *
+ * This class represents a weighted cache.
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/cache/ICache.java]: There is also a cache called `ChunkCache' which also has a weighted type, but it doesn't extend the `ICache' interface. So, we either have to fix javadocs for both of these cases or for none of them. The latter case sounds easier to me as the `CacheSize` already has a description.

File: src/java/org/apache/cassandra/cache/ChunkCache.java [hunk 5]
```diff
@@ -172,7 +171,7 @@ public void onRemoval(Key key, Buffer buffer, RemovalCause cause)
         buffer.release();
     }
 
-    public void close()
+    public void clear()
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/cache/ChunkCache.java:174]: This change is not related to the issue, right?

File: src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java (hunk 3 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java:28]: @Mmuzaf are AuthCaches the only caches ever which will be unweighted? I think the general distinction between weighted and unweighted makes sense. Maybe we could extra document that? Othewise I think it is just ok as it is ...

File: src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+     * Create metrics for given cache.
+     *
+     * @param type  Type of Cache to identify metrics
+     * @param cache Unweighted Cache to measure metrics
+     */
+    public UnweightedCacheMetrics(String type, UnweightedCacheSize cache)
+    {
+        super(new DefaultNameFactory("UnweightedCache", type), type);
+
+        maxEntries = Metrics.register(factory.createMetricName("Capacity"), cache::maxEntries);
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java]: Should we still keep it synced with the `CacheMetrics` names? I think we can use the "maxEntries" instead of the "Capacity" to avoid confusion.

File: src/java/org/apache/cassandra/cache/UnweightedCacheSize.java (hunk 1 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cache/UnweightedCacheSize.java:39]: is not it used in `AuthCache#setMaxEntries`? It basically implements that method which is call elsewhere, like in `SetAuthCacheConfig`.

Comment (smiklosovic) [reply src/java/org/apache/cassandra/cache/UnweightedCacheSize.java:39]: hah so "problem" here is that `AuthCache` implements `AuthCacheMBean` which has a method of the same signature and then it changes the size via `setMaxEntriesDelegate`. I am not going deeper than this. I think this is just fine ... for cases when a cache would not extend MBean with such method name, we would still need a way how to change max entries etc ...

File: src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java (hunk 3 above)
Comment (Mmuzaf) [reply src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java:28]: In the future, however, weighted and unweighted may not be sufficient for us as well, so I wouldn't want to introduce a new concept at this point, as the `AuthCache` fully covers all our needs. It also sounds clearer without adding pages of documentation, and we can also change it in the future if a new type arises and we will decide to classify it... :-)

File: src/java/org/apache/cassandra/tools/NodeProbe.java [hunk 7]
```diff
@@ -1748,22 +1749,22 @@ public Object getCacheMetric(String cacheType, String metricName)
...
                     return JMX.newMBeanProxy(mbeanServerConn,
-                            new ObjectName("org.apache.cassandra.metrics:type=Cache,scope=" + cacheType + ",name=" + metricName),
-                            CassandraMetricsRegistry.JmxTimerMBean.class).getMean();
+                                             new ObjectName("org.apache.cassandra.metrics:type=" + type + ",scope=" + scope + ",name=" + metricName),
+                                             CassandraMetricsRegistry.JmxTimerMBean.class).getMean();
                 case "MissLatencyUnit":
                     return JMX.newMBeanProxy(mbeanServerConn,
-                            new ObjectName("org.apache.cassandra.metrics:type=Cache,scope=" + cacheType + ",name=MissLatency"),
-                            CassandraMetricsRegistry.JmxTimerMBean.class).getDurationUnit();
+                                             new ObjectName("org.apache.cassandra.metrics:type=" + type + ",scope=" + scope + ",name=MissLatency"),
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/tools/NodeProbe.java:1766]: Should we change `name=MissLatency` to `name=MissLatencyUnit` or am I missing something?

File: src/java/org/apache/cassandra/cache/ChunkCache.java (hunk 5 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cache/ChunkCache.java:174]: I cant find any usage of that method. You are technically right, yes. But looking at it, why not to rename it to `clear`?
 The impl invalidates the cache which is kind of a clearing ... I think "clear" makes more sense and reflects what that method does than "close". That cache does not implement `(Auto)Closeable` or anything like that, there is nothing to _close_. I think that was the original idea behind this change. Keep in mind I just put it all together after Sam did his improvements.

File: src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java (hunk 6 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java]: Is this change only internally visible? If it is exposed to clients I think we need to keep it like it is to not break anything.

File: src/java/org/apache/cassandra/tools/NodeProbe.java [hunk 8]
```diff
@@ -1734,10 +1734,11 @@ public long getReadRepairRepairedBackground()
     // JMX getters for the o.a.c.metrics API below.
     /**
      * Retrieve cache metrics based on the cache type (KeyCache, RowCache, or CounterCache)
-     * @param cacheType KeyCach, RowCache, or CounterCache
+     * @param type type, like Cache or UnweightedCache
+     * @param scope KeyCach, RowCache, or CounterCache
      * @param metricName Capacity, Entries, HitRate, Size, Requests or Hits.
      */
-    public Object getCacheMetric(String cacheType, String metricName)
+    public Object getCacheMetric(String type, String scope, String metricName)
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/tools/NodeProbe.java]: Since the `getCacheMetric` method isn't currently used for unweighted caches, would it be better to roll back these changes and do it when the nodetool needs it? 

Another option is to create an overloaded `public Object getCacheMetric` method that takes "Cache" as the first argument, this will help us avoid changes that are not necessary right now. The weighted and unweighted caches are slightly different anyway, so `getCacheMetric` is not quite what we want (e.g. the `size` is different for weighted and unweighted).

Comment (smiklosovic) [reply src/java/org/apache/cassandra/tools/NodeProbe.java]: I think this is leftover from times when I was trying to expose unweighted caches as well into nodetool's `Info` so it was not always called with "Cache" but I removed that functionality from `Info` because I just felt in my guts that this is not right thing to do anyway ... Metrics are going to be exposed via CQL and adding more metrics into nodetool just feels wrong. I can revert this whole change.

File: src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java (hunk 3 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java:28]: I have a hard time to figure out what other metrics we would have so we could not categorize them into weighted or unweighted ones.

File: src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java (hunk 6 above)
Comment (Mmuzaf) [reply src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java]: This will be exposed to an end user, however, as we don't have such metrics in the public API at this moment, we can expose it with the "right" name. Even though auth caches and weighted caches have different MBeans, vtables there is no need to keep the attribute names in sync between them, so I guess we can change it.

File: src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java (hunk 3 above)
Comment (Mmuzaf) [reply src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java:28]: AuthCache metrics are valuable in their own right, though, so I'm not sure we need an additional 'unweighted' category for them. The simpler it is, the better it will be. 
We can move this question to the issue for visibility. 
@beobal I think it was your original idea to organise classes in this way. wdyt ?

Comment (beobal) [reply src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java:28]: I tend to agree with @smiklosovic that the general distinction between weighted and unweighted makes sense.

> In the future, however, weighted and unweighted may not be sufficient for us as well, so I wouldn't want to introduce a new concept at this point, as the AuthCache fully covers all our needs.

I don't think any new concept is being introducing here, unweighted caches exist in C* you are just exposing them via the new interface. 

I also don't agree that this requires pages of additional documentation; _some_ docs will be required to direct users to the appropriate metrics, so it seems to me a choice between 

`"For metrics related to auth caches, see the AuthCache metrics 'nameX, nameY, nameZ...'"`
vs
`"For metrics related auth caches, see the UnweightedCache metrics 'nameX, nameY, nameZ...'"`

I don't believe that docs for the metrics themselves don't need to go into great detail about the nature of weighted vs unweighted caches.

That's my $0.02, but I don't feel mega-strongly about it so wouldn't object whichever way you decide.

File: src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java (hunk 6 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java]: @Mmuzaf I would rather stay consistent and rename `maxEntries` gauge to `capacity` to be on par with `CacheMetrics` which were there first which have it `capacity` too. There is no reason to have this naming discrepancy among different types of cache metrics and term "capacity" is just perfectly fine.

Comment (smiklosovic) [reply src/java/org/apache/cassandra/metrics/UnweightedCacheMetrics.java]: you know what ... you are right. `Capacity` here should be `MaxEntries`. Because `Capacity` and `MaxEntries` are not the same thing.

File: doc/modules/cassandra/pages/managing/operating/metrics.adoc [hunk 9]
```diff
@@ -528,17 +528,17 @@ ViewReplicasSuccess.
 base table and when CL.ONE is achieved on view.
 |===
 
-== Cache Metrics
+== Weighted Cache Metrics
```
Comment (Mmuzaf) [inline doc/modules/cassandra/pages/managing/operating/metrics.adoc:531]: Should we also add a few words about 'weighted' and 'unweighted' terms? I think here is the comment with a well-described difference, we can use it:

https://issues.apache.org/jira/browse/CASSANDRA-17062?focusedCommentId=17443143&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17443143

File: test/unit/org/apache/cassandra/auth/AuthCacheTest.java [hunk 10]
```diff
@@ -322,6 +322,41 @@ public void testCacheLoaderIsNotCalledOnGetAllWhenCacheIsNotEmpty()
         assertEquals(1, loadCounter);
     }
 
+    @Test
+    public void testMetricsOnCacheEnabled()
```
Comment (Mmuzaf) [inline test/unit/org/apache/cassandra/auth/AuthCacheTest.java:335]: Should we also check the numbers for the `entries` and `maxEntries` method calls? They do not seem to be covered by the tests.

File: test/unit/org/apache/cassandra/auth/AuthCacheTest.java [hunk 11]
```diff
@@ -322,6 +322,41 @@ public void testCacheLoaderIsNotCalledOnGetAllWhenCacheIsNotEmpty()
...
+        authCache.get("11");
+
+        assertThat(authCache.getMetrics().requests.getCount()).isEqualTo(4L);
+        assertThat(authCache.getMetrics().hits.getCount()).isEqualTo(2L);
+        assertThat(authCache.getMetrics().misses.getCount()).isEqualTo(2L);
+        assertEquals(2, loadCounter);
+    }
+
+    @Test
+    public void testMetricsOnCacheDisabled()
```
Comment (Mmuzaf) [inline test/unit/org/apache/cassandra/auth/AuthCacheTest.java:376]: Should we add a test to check that new metrics are available via the JMX API? I haven't found a good candidate for a test class, but I think it's still worth checking that `getMaxEntries` and `setMaxEntries` match, and the `entries` match the number entries we set to the cache.

File: src/java/org/apache/cassandra/auth/AuthCache.java [hunk 12]
```diff
@@ -225,6 +235,7 @@ public V get(K k)
         if (cache == null)
             return loadFunction.apply(k);
 
+        metrics.requests.mark();
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/auth/AuthCache.java]: Aren't we missing the `getAll` path for tracking requests here? The `ChunkCache` itself encapsulates this logic in the `StatsCounter`.  It's worth adding such a test or a verification the metrics value for existing tests in `AuthCacheTest`.

File: src/java/org/apache/cassandra/cache/UnweightedCacheSize.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Represents an unweighted cache size.
+ */
+public interface UnweightedCacheSize
+{
+    /**
+     * Returns the maximum number of entries in this cache.
+     */
+    int maxEntries();
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/cache/UnweightedCacheSize.java:34]: Should this be renamed to `getMaxEntries`?

File: src/java/org/apache/cassandra/auth/AuthCache.java [hunk 14]
```diff
@@ -52,12 +58,13 @@
...
 import static org.apache.cassandra.config.CassandraRelevantProperties.DISABLE_AUTH_CACHES_REMOTE_CONFIGURATION;
 
-public class AuthCache<K, V> implements AuthCacheMBean, Shutdownable
+public class AuthCache<K, V> implements AuthCacheMBean, UnweightedCacheSize, Shutdownable
 {
     private static final Logger logger = LoggerFactory.getLogger(AuthCache.class);
 
     public static final String MBEAN_NAME_BASE = "org.apache.cassandra.auth:type=";
 
+    @SuppressWarnings("rawtypes")
```
Comment (Mmuzaf) [inline src/java/org/apache/cassandra/auth/AuthCache.java]: Maybe we should use `ScheduledFuture<?>` instead?

File: doc/modules/cassandra/pages/managing/operating/metrics.adoc (hunk 9 above)
Comment (smiklosovic) [reply doc/modules/cassandra/pages/managing/operating/metrics.adoc:531]: done

File: src/java/org/apache/cassandra/auth/AuthCache.java (hunk 14 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/auth/AuthCache.java]: huh how did this happen ... I don't remember putting it there, sure, can be like suggested.

File: src/java/org/apache/cassandra/cache/UnweightedCacheSize.java (hunk 13 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/cache/UnweightedCacheSize.java:34]: The naming of these methods basically copies these in `CacheSize`. I would keep it like it is. In `CacheSize` there is `capacity` / `getCapacity` etc ... We would need to rework everything, `int size()`, `long weightedSize()` to prefix with `get` etc ...

File: src/java/org/apache/cassandra/auth/AuthCache.java (hunk 12 above)
Comment (smiklosovic) [reply src/java/org/apache/cassandra/auth/AuthCache.java]: Hmm ... good catch.

File: test/unit/org/apache/cassandra/auth/AuthCacheTest.java (hunk 10 above)
Comment (smiklosovic) [reply test/unit/org/apache/cassandra/auth/AuthCacheTest.java:335]: done

File: test/unit/org/apache/cassandra/auth/AuthCacheTest.java (hunk 11 above)
Comment (smiklosovic) [reply test/unit/org/apache/cassandra/auth/AuthCacheTest.java:376]: I dont think that involving _JMX_ is necessary. We basically just need to invoke same MBean method programmatically. I have added "testSettingMaxEntries" test for that.

**Review by Mmuzaf — APPROVED** [review]
LGTM, +1
